### ISSUE TITLE
bug: InvocationTargetException when a GraphQLError is thrown for a federated entity query

### ISSUE THREAD

**Issue by marcelvanderperk** [Netflix/dgs-framework#229]
## Expected behavior
That my GraphQLError is returned to the client.

## Actual behavior
The GraphQLError gets wrapped in a InvocationTargetException and hides the real exception from the client. I have written a workaround to unwrap this exception in a DataFetcherExceptionHandler, but I think this should work for everyone without having to write a custom exception handler.

## Steps to reproduce

I'm using the latest version 3.10.2 and its easily reproducible with this basic example; 

```java
  @DgsEntityFetcher(name = "Image")
  public Image getImageEntity(Map<String, Object> values) {
    throw new ImageDoesNotExist();
  }

  public static class ImageDoesNotExist extends RuntimeException implements GraphQLError {

    @Override
    public String getMessage() {
      return "Blabla";
    }

    @Override
    public List<SourceLocation> getLocations() {
      return Collections.emptyList();
    }

    @Override
    public ErrorClassification getErrorType() {
      return ErrorType.NOT_FOUND;
    }
  }
```

The Query
```json
query ($representations: [_Any!]!) {
    _entities(representations: $representations) {
        ... on Image {
            id 
            url
        }
    }
}
```

GraphQL variable

```json
{
    "representations": [
        {
            "__typename": "Image",
            "id": "34b85b92-e219-4fbb-8bf2-682c4172682b"
        }
    ]
}
```

This results in.
```json
{
    "errors": [
        {
            "message": "java.lang.reflect.InvocationTargetException: null",
            "locations": [],
            "path": [
                "_entities"
            ],
            "extensions": {
                "errorType": "INTERNAL"
            }
        }
    ],
    "data": null
}
```

My current workaround by unwrapping the InvocationTargetException in a custom DataFetcherExceptionHandler.

```java
@Component
public class DgsExceptionHandler implements DataFetcherExceptionHandler {
  private final DefaultDataFetcherExceptionHandler defaultHandler = new DefaultDataFetcherExceptionHandler();

  @Override
  public DataFetcherExceptionHandlerResult onException(DataFetcherExceptionHandlerParameters handlerParameters) {
    //exceptions seem to get wrapped in a InvocationTargetException
    if (handlerParameters.getException() instanceof InvocationTargetException) {
      InvocationTargetException invocationTargetException = (InvocationTargetException) handlerParameters.getException();
      Throwable targetException = invocationTargetException.getTargetException();
      if (targetException instanceof GraphQLError) {
        return DataFetcherExceptionHandlerResult.newResult()
            .error((GraphQLError) targetException)
            .build();
      }
    }
    return defaultHandler.onException(handlerParameters);
  }
}
```

```json
{
    "errors": [
        {
            "message": "Blabla",
            "locations": [],
            "extensions": {
                "classification": "BAD_REQUEST"
            }
        },
        {
            "message": "The field at path '/_entities' was declared as a non null type, but the code involved in retrieving data has wrongly returned a null value.  The graphql specification requires that the parent field be set to null, or if that is non nullable that it bubble up null to its parent and so on. The non-nullable type is '[_Entity]' within parent type 'Query'",
            "path": [
                "_entities"
            ],
            "extensions": {
                "classification": "NullValueInNonNullableField"
            }
        }
    ],
    "data": null
}
```

**Comment by srinivasankavitha** [comment]
Thanks for reporting. We’ll look into it.

**Comment by srinivasankavitha** [comment]
So, we handle some standard exceptions by populating the error information in the response. Since in your example, you are actually creating a custom exception, you will need to write an exception handler for it. It is easier than how you implemented it though, because you don't need to check for InvocationTargetException. Rather, just check for the specific custom exception that you are throwing.

```
@Component
public class CustomDataFetchingExceptionHandler implements DataFetcherExceptionHandler {
    private final DefaultDataFetcherExceptionHandler defaultHandler = new DefaultDataFetcherExceptionHandler();

    @Override
    public DataFetcherExceptionHandlerResult onException(DataFetcherExceptionHandlerParameters handlerParameters) {
        if(handlerParameters.getException() instanceof ImageDoesNotExistException) {
            Map<String, Object> debugInfo = new HashMap<>();
            debugInfo.put("somefield", "somevalue");

            GraphQLError graphqlError = TypedGraphQLError.INTERNAL.message("Image does not exist!")
                    .debugInfo(debugInfo)
                    .path(handlerParameters.getPath()).build();
            return DataFetcherExceptionHandlerResult.newResult()
                    .error(graphqlError)
                    .build();
        } else {
            return defaultHandler.onException(handlerParameters);
        }
    }
}
```

https://netflix.github.io/dgs/error-handling/

It is not possible for us to handle custom exceptions in a standard way because the error type would differ based on the type of exception.
Hope that helps.

**Comment by marcelvanderperk** [comment]
I just executed it again and I still have the same issue. I get this stacktrace (should have included that in the original report).
I'd like to emphasize this only happens on a federated entity query. 

```
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.netflix.graphql.dgs.federation.DefaultDgsFederationResolver.dgsEntityFetchers(DefaultDgsFederationResolver.kt:78)
	at com.netflix.graphql.dgs.federation.DefaultDgsFederationResolver.access$dgsEntityFetchers(DefaultDgsFederationResolver.kt:36)
	at com.netflix.graphql.dgs.federation.DefaultDgsFederationResolver$entitiesFetcher$1.get(DefaultDgsFederationResolver.kt:59)
	at com.netflix.graphql.dgs.metrics.micrometer.DgsGraphQLMetricsInstrumentation$instrumentDataFetcher$1.get(DgsGraphQLMetricsInstrumentation.kt:91)
	at graphql.execution.ExecutionStrategy.fetchField(ExecutionStrategy.java:270)
	at graphql.execution.ExecutionStrategy.resolveFieldWithInfo(ExecutionStrategy.java:203)
	at graphql.execution.AsyncExecutionStrategy.execute(AsyncExecutionStrategy.java:60)
	at graphql.execution.Execution.executeOperation(Execution.java:165)
	at graphql.execution.Execution.execute(Execution.java:104)
	at graphql.GraphQL.execute(GraphQL.java:557)
	at graphql.GraphQL.parseValidateAndExecute(GraphQL.java:482)
	at graphql.GraphQL.executeAsync(GraphQL.java:446)
	at graphql.GraphQL.execute(GraphQL.java:377)
	at com.netflix.graphql.dgs.internal.DefaultDgsQueryExecutor.execute(DefaultDgsQueryExecutor.kt:114)
	at com.netflix.graphql.dgs.internal.DefaultDgsQueryExecutor.execute(DefaultDgsQueryExecutor.kt:133)
	at com.netflix.graphql.dgs.DgsQueryExecutor.execute(DgsQueryExecutor.kt:37)
	at com.netflix.graphql.dgs.internal.DefaultDgsQueryExecutor.getJsonResult(DefaultDgsQueryExecutor.kt:196)
	at com.netflix.graphql.dgs.internal.DefaultDgsQueryExecutor.executeAndGetDocumentContext(DefaultDgsQueryExecutor.kt:192)
```

Your suggestion of `if(handlerParameters.getException() instanceof ImageDoesNotExistException) ` does not work because my exception does not extend `InvocationTargetException` in any way and it resolves to false.

**Comment by srinivasankavitha** [comment]
I see, ok, that information helps. I'll look into reproducing it with the entity query.

**Comment by srinivasankavitha** [comment]
Implemented a fix, will be available in the next release this week: #253
